src/databricks/labs/ucx/progress/history.py

```diff
@@ -0,0 +1,216 @@
+        id_attribute_names = tuple(klazz.__id_attributes__)
+        all_fields = self._field_types
+        for name in id_attribute_names:
+            id_attribute_type = all_fields.get(name, None) or self._detect_property_type(klazz, name)
+            if id_attribute_type is None:
+                raise AttributeError(name=name, obj=klazz)
+            if id_attribute_type != str:
+                msg = f"Historical record {klazz} has a non-string id attribute: {name} (type={id_attribute_type})"
+                raise TypeError(msg)
+        return id_attribute_names
+
+    def _detect_property_type(self, klazz: type[Record], name: str) -> str | None:
+        maybe_property = getattr(klazz, name, None)
+        if maybe_property is None:
+            return None
+        if not isinstance(maybe_property, property):
+            msg = f"Historical record {klazz} declares an id attribute that is not a field or property: {name} (type={maybe_property})"
+            raise TypeError(msg)
+        property_getter = maybe_property.fget
+        if not property_getter:
+            msg = f"Historical record {klazz} has a non-readable property as an id attribute: {name}"
+            raise TypeError(msg)
+        type_hints = get_type_hints(property_getter) or {}
+        try:
+            return type_hints["return"]
+        except KeyError as e:
+            msg = f"Historical record {klazz} has a property with no type as an id attribute: {name}"
+            raise TypeError(msg) from e
+
+    @classmethod
+    def _get_object_type(cls, klass: type[Record]) -> str:
+        return klass.__name__
+
+    @classmethod
+    def _as_dict(cls, record: Record) -> dict[str, Any]:
+        return dataclasses.asdict(record)
+
+    def _object_id(self, record: Record) -> list[str]:
+        return [getattr(record, field) for field in self._id_attribute_names]
+
+    @classmethod
+    def _encode_non_serializable(cls, name: str, value: Any) -> Any:
+        if isinstance(value, dt.datetime):
+            # Only allow tz-aware timestamps.
+            if value.tzinfo is None:
+                # Name refers to the outermost field, not necessarily a field on a (nested) dataclass.
+                msg = f"Timestamp without timezone not supported in or within field {name}: {value}"
+                raise ValueError(msg)
+            # Always store with 'Z'.
+            ts_utc = value.astimezone(dt.timezone.utc)
+            return ts_utc.isoformat().replace("+00:00", "Z")
+
+        msg = f"Cannot encode {type(value)} value in or within field {name}: {value!r}"
+        raise TypeError(msg)
+
+    def _encode_field_value(self, name: str, value: Any | None) -> str | None:
+        if value is None:
+            return None
+        value_type = self._field_types[name]
+        if value_type in (str, (str | None)):
+            return value
+        encoded_value = json.dumps(
+            value,
+            allow_nan=False,
+            separators=(",", ":"),
```

Comment:
Add a comment why this: no space behind default separator. (I had to Google this)